processing/src/main/java/io/druid/query/topn/TopNBinaryFn.java

```diff
@@ -136,4 +183,82 @@ public TopNBinaryFn(
     }
     return bob.build();
   }
+
+  private DimensionAndMetricValueExtractor getExtractorWithOverriddenValue(
+      String dimension,
+      Object overrideVal,
+      DimensionAndMetricValueExtractor originalExtractor
+  )
+  {
+    Map<String, Object> newMap = Maps.newHashMap(originalExtractor.getBaseObject());
+    newMap.put(dimension, overrideVal);
+    DimensionAndMetricValueExtractor newExtractor = new DimensionAndMetricValueExtractor(newMap);
+    return newExtractor;
+  }
+
+  private Function<Object, Object> initValueTransformer()
+  {
+    if (typeHints == null) {
+      return STRING_TRANSFORMER;
+    }
+
+    ValueType type = typeHints.get(dimension);
+    if (type == null) {
+      type = ValueType.STRING;
+    }
+    switch (type) {
+      case STRING:
+        return STRING_TRANSFORMER;
+      case LONG:
+        return LONG_TRANSFORMER;
+      case FLOAT:
+        return FLOAT_TRANSFORMER;
+      default:
+        throw new IAE("invalid type");
+    }
+  }
+
+  private DimensionAndMetricValueExtractor getTransformedExtractor(DimensionAndMetricValueExtractor argVal)
```

Comment:
I moved the transform functions to `TopNMapFn`, the transformation is now applied within `DimExtractionTopNAlgorithm` as it adds results to the result builder